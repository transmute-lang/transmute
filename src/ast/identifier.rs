use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, PartialEq)]
pub struct Identifier {
    name: String,
    span: Span,
}

impl Identifier {
    pub fn new(name: String, span: Span) -> Self {
        Self { name, span }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}

impl Display for Identifier {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}
