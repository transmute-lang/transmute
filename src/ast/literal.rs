use crate::ast::ids::IdentRefId;
use crate::lexer::Span;
use std::fmt::{Debug, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    kind: LiteralKind,
    span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }

    pub fn kind(&self) -> &LiteralKind {
        &self.kind
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
