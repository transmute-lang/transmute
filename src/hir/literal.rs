use crate::ast::literal::Literal as AstLiteral;
use crate::ast::literal::LiteralKind as AstLiteralKind;
use crate::ids::IdentRefId;
use crate::lexer::Span;
use std::fmt::{Debug, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    kind: LiteralKind,
    span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }

    pub fn kind(&self) -> &LiteralKind {
        &self.kind
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

impl From<AstLiteral> for Literal {
    fn from(value: AstLiteral) -> Self {
        Self {
            span: value.span().clone(),
            kind: LiteralKind::from(value.take_kind()),
        }
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl From<AstLiteralKind> for LiteralKind {
    fn from(value: AstLiteralKind) -> Self {
        match value {
            AstLiteralKind::Boolean(b) => Self::Boolean(b),
            AstLiteralKind::Identifier(i) => Self::Identifier(i),
            AstLiteralKind::Number(n) => Self::Number(n),
        }
    }
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
