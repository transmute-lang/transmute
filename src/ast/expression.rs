use crate::ast::literal::Literal;
use crate::ast::operators::{BinaryOperator, UnaryOperator};
use crate::ast::Visitor;
use crate::lexer::{Location, Span};

#[derive(Debug, PartialEq)]
pub struct Expression {
    kind: ExpressionKind,
}

impl Expression {
    pub fn new(kind: ExpressionKind) -> Self {
        Self { kind }
    }

    pub fn kind(&self) -> &ExpressionKind {
        &self.kind
    }

    pub fn location(&self) -> &Location {
        match &self.kind {
            ExpressionKind::Literal(l) => l.location(),
            ExpressionKind::Binary(e, _, _) => e.location(),
            ExpressionKind::Unary(o, _) => o.location(),
        }
    }

    pub fn span(&self) -> Span {
        match &self.kind {
            ExpressionKind::Literal(l) => l.span().clone(),
            ExpressionKind::Binary(l, _, r) => l.span().extend_to(&r.span()),
            ExpressionKind::Unary(o, e) => o.span().extend_to(&e.span()),
        }
    }

    pub fn accept<V, R>(&self, visitor: &mut V) -> R
    where
        V: Visitor<R>,
    {
        visitor.visit_expression(self)
    }
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Self {
            kind: ExpressionKind::Literal(literal),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum ExpressionKind {
    Literal(Literal),
    Binary(Box<Expression>, BinaryOperator, Box<Expression>),
    Unary(UnaryOperator, Box<Expression>),
}
