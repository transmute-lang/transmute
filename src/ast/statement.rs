use crate::ast::expression::ExprId;
use crate::ast::identifier::Identifier;
use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, PartialEq)]
pub struct Statement<'s> {
    id: StmtId,
    // todo should be merged with Expression, maybe?
    kind: StatementKind<'s>,
    span: Span,
}

impl<'s> Statement<'s> {
    pub fn new(id: StmtId, kind: StatementKind<'s>, span: Span) -> Self {
        Self { id, kind, span }
    }

    pub fn id(&self) -> StmtId {
        self.id
    }

    pub fn kind(&self) -> &StatementKind<'s> {
        &self.kind
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct StmtId {
    id: usize,
}

impl StmtId {
    pub fn id(&self) -> usize {
        self.id
    }
}

impl Display for StmtId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.id)
    }
}

impl From<usize> for StmtId {
    fn from(id: usize) -> Self {
        Self { id }
    }
}

#[derive(Debug, PartialEq)]
pub enum StatementKind<'s> {
    Expression(ExprId),
    Let(Identifier<'s>, ExprId),
    Ret(ExprId),
    // todo vec does not hold span and position as it should (search for "Vec<Statement>, Span")
    LetFn(Identifier<'s>, Vec<Identifier<'s>>, Vec<StmtId>),
}
