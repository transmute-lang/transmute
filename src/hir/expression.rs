use crate::ast::expression::Expression as AstExpression;
use crate::ast::expression::ExpressionKind as AstExpressionKind;
use crate::ast::expression::Target as AstTarget;
use crate::hir::literal::Literal;
use crate::hir::typed::{Typed, TypedState, Untyped};
use crate::ids::{ExprId, IdentRefId, StmtId, TypeId};
use crate::lexer::Span;
use std::fmt::Debug;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression<T>
where
    T: TypedState,
{
    id: ExprId,
    kind: ExpressionKind,
    span: Span,
    typed: T,
}

impl Expression<Untyped> {
    pub fn typed(self, ty: TypeId) -> Expression<Typed> {
        Expression {
            id: self.id,
            kind: self.kind,
            span: self.span,
            typed: Typed(ty),
        }
    }
}

impl From<AstExpression> for Expression<Untyped> {
    fn from(value: AstExpression) -> Self {
        Self {
            id: value.id(),
            span: value.span().clone(),
            kind: match value.take_kind() {
                AstExpressionKind::Assignment(target, expr_id) => {
                    ExpressionKind::Assignment(Target::from(target), expr_id)
                }
                AstExpressionKind::If(cond, true_branch, false_branch) => {
                    ExpressionKind::If(cond, true_branch, false_branch)
                }
                AstExpressionKind::Literal(literal) => {
                    ExpressionKind::Literal(Literal::from(literal))
                }
                AstExpressionKind::Access(expr_id, ident_ref_id) => {
                    ExpressionKind::Access(expr_id, ident_ref_id)
                }
                AstExpressionKind::FunctionCall(ident_ref_id, expr_ids) => {
                    ExpressionKind::FunctionCall(ident_ref_id, expr_ids)
                }
                AstExpressionKind::While(cond, expr_id) => ExpressionKind::While(cond, expr_id),
                AstExpressionKind::Block(expr_ids) => ExpressionKind::Block(expr_ids),
                AstExpressionKind::StructInstantiation(ident_ref_id, fields) => {
                    ExpressionKind::StructInstantiation(ident_ref_id, fields)
                }
                AstExpressionKind::Unary(_, _) => panic!("Cannot convert AstExpressionKind::Unary"),
                AstExpressionKind::Binary(_, _, _) => {
                    panic!("Cannot convert AstExpressionKind::Binary")
                }
                AstExpressionKind::Dummy => panic!("Cannot convert AstExpressionKind::Dummy"),
            },
            typed: Untyped,
        }
    }
}

impl<T> Expression<T>
where
    T: TypedState,
{
    pub fn id(&self) -> ExprId {
        self.id
    }

    pub fn kind(&self) -> &ExpressionKind {
        &self.kind
    }

    pub fn set_span(&mut self, span: Span) {
        self.span = span;
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

impl Expression<Typed> {
    pub fn type_id(&self) -> TypeId {
        self.typed.0
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}

impl From<AstTarget> for Target {
    fn from(value: AstTarget) -> Self {
        match value {
            AstTarget::Direct(ident_ref_id) => Self::Direct(ident_ref_id),
            AstTarget::Indirect(expr_id) => Self::Indirect(expr_id),
        }
    }
}
