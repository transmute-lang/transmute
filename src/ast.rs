pub mod expression;
pub mod identifier;
pub mod literal;
pub mod operators;
pub mod statement;

use crate::ast::expression::{ExprId, Expression, ExpressionKind};
use crate::ast::identifier::IdentId;
use crate::ast::literal::{Literal, LiteralKind};
use crate::ast::statement::{Statement, StatementKind, StmtId};
use std::fmt::{Display, Formatter};

pub trait Visitor<R> {
    fn visit_ast(&mut self, ast: &Ast) -> R;

    fn visit_statement(&mut self, stmt: StmtId) -> R;

    fn visit_expression(&mut self, expr: ExprId) -> R;

    fn visit_literal(&mut self, literal: &Literal) -> R;
}

#[derive(Debug, PartialEq)]
pub struct Ast {
    identifiers: Vec<String>,
    expressions: Vec<Expression>,
    statements: Vec<Statement>,
    // todo replace with Statements
    root: Vec<StmtId>,
}

impl Ast {
    pub fn new(
        identifiers: Vec<String>,
        expressions: Vec<Expression>,
        statements: Vec<Statement>,
        root: Vec<StmtId>,
    ) -> Self {
        assert!(!root.is_empty());
        Self {
            identifiers,
            expressions,
            statements,
            root,
        }
    }

    pub fn statements(&self) -> &Vec<StmtId> {
        &self.root
    }

    pub fn identifier(&self, id: IdentId) -> &String {
        &self.identifiers[id.id()]
    }

    pub fn expression(&self, id: ExprId) -> &Expression {
        &self.expressions[id.id()]
    }

    pub fn statement(&self, id: StmtId) -> &Statement {
        &self.statements[id.id()]
    }
}

pub struct PrettyPrint<'a, T> {
    indent: usize,
    ast: &'a Ast,
    id: T,
}

impl<'a, T> PrettyPrint<'a, T> {
    pub fn new(ast: &'a Ast, id: T) -> Self {
        Self { indent: 0, ast, id }
    }

    fn new_with_ident(ast: &'a Ast, id: T, ident: usize) -> Self {
        Self {
            indent: ident,
            ast,
            id,
        }
    }
}

impl Display for PrettyPrint<'_, StmtId> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let indent = "  ".repeat(self.indent);
        match self.ast.statement(self.id).kind() {
            StatementKind::Expression(expr) => {
                let semi = match self.ast.expression(*expr).kind() {
                    ExpressionKind::If(_, _, _) | ExpressionKind::While(_, _)=> "",
                    _ => ";",
                };
                writeln!(
                    f,
                    "{indent}{}{semi}",
                    PrettyPrint::new_with_ident(self.ast, *expr, self.indent)
                )
            }
            StatementKind::Let(ident, expr) => {
                writeln!(
                    f,
                    "{indent}let {} = {};",
                    PrettyPrint::new(self.ast, ident.id()),
                    PrettyPrint::new(self.ast, *expr),
                )
            }
            StatementKind::Ret(expr) => {
                writeln!(f, "{indent}ret {};", PrettyPrint::new(self.ast, *expr))
            }
            StatementKind::LetFn(ident, params, stmts) => {
                write!(f, "{indent}let {}(", PrettyPrint::new(self.ast, ident.id()))?;
                for (i, param) in params.iter().enumerate() {
                    write!(f, "{}", PrettyPrint::new(self.ast, param.id()))?;
                    if i < params.len() - 1 {
                        write!(f, ", ")?;
                    }
                }
                writeln!(f, ") = {{")?;
                for stmt in stmts {
                    write!(
                        f,
                        "{indent}{}",
                        PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                    )?;
                }
                writeln!(f, "{indent}}}")
            }
        }
    }
}

impl Display for PrettyPrint<'_, ExprId> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let indent = "  ".repeat(self.indent);
        match self.ast.expression(self.id).kind() {
            ExpressionKind::Assignment(ident, expr) => {
                write!(
                    f,
                    "{} = {}",
                    PrettyPrint::new(self.ast, ident.id()),
                    PrettyPrint::new(self.ast, *expr)
                )
            }
            ExpressionKind::If(cond, true_branch, false_branch) => {
                writeln!(f, "if {} {{", PrettyPrint::new(self.ast, *cond))?;
                for stmt in true_branch {
                    write!(
                        f,
                        "{}",
                        PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                    )?;
                }
                if !false_branch.is_empty() {
                    writeln!(f, "}}")?;
                    writeln!(f, "else {{")?;
                    for stmt in false_branch {
                        write!(
                            f,
                            "{}",
                            PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                        )?;
                    }
                }
                write!(f, "{indent}}}")
            }
            ExpressionKind::Literal(lit) => match lit.kind() {
                LiteralKind::Boolean(b) => write!(f, "{b}"),
                LiteralKind::Identifier(ident) => {
                    write!(f, "{}", PrettyPrint::new(self.ast, ident.id()))
                }
                LiteralKind::Number(n) => write!(f, "{n}"),
            },
            ExpressionKind::Binary(left, op, right) => {
                write!(
                    f,
                    "{} {} {}",
                    PrettyPrint::new(self.ast, *left),
                    op.kind(),
                    PrettyPrint::new(self.ast, *right)
                )
            }
            ExpressionKind::FunctionCall(ident, params) => {
                write!(f, "{}(", PrettyPrint::new(self.ast, ident.id()))?;
                for (i, expr) in params.iter().enumerate() {
                    write!(f, "{}", PrettyPrint::new(self.ast, *expr))?;
                    if i < params.len() - 1 {
                        write!(f, ", ")?;
                    }
                }
                write!(f, ")")
            }
            ExpressionKind::Unary(op, expr) => {
                write!(f, "{} {}", op.kind(), PrettyPrint::new(self.ast, *expr))
            }
            ExpressionKind::While(cond, stmts) => {
                writeln!(f, "while {} {{", PrettyPrint::new(self.ast, *cond))?;
                for stmt in stmts {
                    write!(
                        f,
                        "{}",
                        PrettyPrint::new_with_ident(self.ast, *stmt, self.indent + 1)
                    )?;
                }
                writeln!(f, "}}")
            }
        }
    }
}

impl Display for PrettyPrint<'_, IdentId> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let indent = "  ".repeat(self.indent);
        write!(f, "{indent}{}", self.ast.identifier(self.id))
    }
}
