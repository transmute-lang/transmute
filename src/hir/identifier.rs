use crate::hir::bound::{Bound, BoundState, Unbound};
use crate::ids::{IdentId, SymbolId};
use crate::lexer::Span;

use crate::ast::identifier::Identifier as AstIdentifier;

/// Represents an identifier when not used as a reference
#[derive(Debug, Clone, PartialEq)]
pub struct Identifier<B>
where
    B: BoundState,
{
    id: IdentId,
    span: Span,
    bound: B,
}

impl Identifier<Unbound> {
    pub fn bind(self, symbol_id: SymbolId) -> Identifier<Bound> {
        Identifier::<Bound> {
            id: self.id,
            span: self.span,
            bound: Bound(symbol_id),
        }
    }
}

impl From<AstIdentifier> for Identifier<Unbound> {
    fn from(value: AstIdentifier) -> Self {
        Identifier {
            id: value.id(),
            span: value.take_span(),
            bound: Unbound,
        }
    }
}

impl Identifier<Bound> {
    pub fn symbol_id(&self) -> SymbolId {
        self.bound.0
    }
}

impl<B> Identifier<B>
where
    B: BoundState,
{
    pub fn id(&self) -> IdentId {
        self.id
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}
