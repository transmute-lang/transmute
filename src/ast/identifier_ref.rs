use crate::ast::identifier::Identifier;
use crate::ast::ids::{IdentId, IdentRefId, SymbolId};
use crate::lexer::Span;

/// Represents an identifier when used as a reference
#[derive(Debug, Clone, PartialEq)]
pub struct IdentifierRef<B>
where
    B: BoundState,
{
    /// ID of this identifier reference
    id: IdentRefId,
    /// The referenced symbol identifier
    ident: Identifier,
    /// When `Bound`, the referenced `SymbolId`
    bound_state: B, // todo move to Identifier
}

impl<S> IdentifierRef<S>
where
    S: BoundState,
{
    pub fn id(&self) -> IdentRefId {
        self.id
    }

    pub fn ident(&self) -> &Identifier {
        &self.ident
    }

    pub fn ident_id(&self) -> IdentId {
        self.ident.id()
    }

    pub fn span(&self) -> &Span {
        self.ident.span()
    }
}

impl IdentifierRef<Unbound> {
    pub fn new(id: IdentRefId, ident: Identifier) -> Self {
        Self {
            id,
            ident,
            bound_state: Unbound {},
        }
    }

    pub fn resolved(self, symbol_id: SymbolId) -> IdentifierRef<Bound> {
        IdentifierRef {
            id: self.id,
            ident: self.ident,
            bound_state: Bound(symbol_id),
        }
    }
}

impl IdentifierRef<Bound> {
    pub fn new_resolved(id: IdentRefId, ident: Identifier, symbol_id: SymbolId) -> Self {
        Self {
            id,
            ident,
            bound_state: Bound(symbol_id),
        }
    }
    pub fn symbol_id(&self) -> SymbolId {
        self.bound_state.0
    }
}

pub trait BoundState {}

#[derive(Debug, Clone, PartialEq)]
pub struct Unbound;

impl BoundState for Unbound {}

#[derive(Debug, Clone, PartialEq)]
pub struct Bound(pub SymbolId);

impl BoundState for Bound {}
