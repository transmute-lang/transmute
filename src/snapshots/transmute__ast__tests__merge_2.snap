---
source: src/ast.rs
expression: "&ast1.merge(ast2)"
---
Ast {
    identifiers: [
        "x_1",
        "x_2",
    ],
    identifier_refs: [
        IdentifierRef {
            id: IdentRefId {
                id: 0,
            },
            ident: Identifier {
                id: IdentId {
                    id: 0,
                },
                span: 1:21; [20, 23],
            },
            bound_state: Unbound,
        },
        IdentifierRef {
            id: IdentRefId {
                id: 1,
            },
            ident: Identifier {
                id: IdentId {
                    id: 0,
                },
                span: 1:14; [13, 16],
            },
            bound_state: Unbound,
        },
        IdentifierRef {
            id: IdentRefId {
                id: 2,
            },
            ident: Identifier {
                id: IdentId {
                    id: 1,
                },
                span: 1:21; [20, 23],
            },
            bound_state: Unbound,
        },
        IdentifierRef {
            id: IdentRefId {
                id: 3,
            },
            ident: Identifier {
                id: IdentId {
                    id: 1,
                },
                span: 1:14; [13, 16],
            },
            bound_state: Unbound,
        },
    ],
    expressions: [
        Expression {
            id: ExprId {
                id: 0,
            },
            kind: Literal(
                Literal {
                    kind: Number(
                        0,
                    ),
                    span: 1:11; [10, 11],
                },
            ),
            span: 1:11; [10, 11],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 1,
            },
            kind: Literal(
                Literal {
                    kind: Identifier(
                        IdentRefId {
                            id: 0,
                        },
                    ),
                    span: 1:21; [20, 23],
                },
            ),
            span: 1:21; [20, 23],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 2,
            },
            kind: Unary(
                UnaryOperator {
                    kind: Minus,
                    span: 1:20; [19, 20],
                },
                ExprId {
                    id: 1,
                },
            ),
            span: 1:20; [19, 23],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 3,
            },
            kind: Literal(
                Literal {
                    kind: Number(
                        2,
                    ),
                    span: 1:27; [26, 27],
                },
            ),
            span: 1:27; [26, 27],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 4,
            },
            kind: Binary(
                ExprId {
                    id: 2,
                },
                BinaryOperator {
                    kind: Multiplication,
                    span: 1:25; [24, 25],
                },
                ExprId {
                    id: 3,
                },
            ),
            span: 1:20; [19, 27],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 5,
            },
            kind: Assignment(
                IdentRefId {
                    id: 1,
                },
                ExprId {
                    id: 4,
                },
            ),
            span: 1:14; [13, 27],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 6,
            },
            kind: Literal(
                Literal {
                    kind: Number(
                        0,
                    ),
                    span: 1:11; [10, 11],
                },
            ),
            span: 1:11; [10, 11],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 7,
            },
            kind: Literal(
                Literal {
                    kind: Identifier(
                        IdentRefId {
                            id: 2,
                        },
                    ),
                    span: 1:21; [20, 23],
                },
            ),
            span: 1:21; [20, 23],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 8,
            },
            kind: Unary(
                UnaryOperator {
                    kind: Minus,
                    span: 1:20; [19, 20],
                },
                ExprId {
                    id: 7,
                },
            ),
            span: 1:20; [19, 23],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 9,
            },
            kind: Literal(
                Literal {
                    kind: Number(
                        2,
                    ),
                    span: 1:27; [26, 27],
                },
            ),
            span: 1:27; [26, 27],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 10,
            },
            kind: Binary(
                ExprId {
                    id: 8,
                },
                BinaryOperator {
                    kind: Multiplication,
                    span: 1:25; [24, 25],
                },
                ExprId {
                    id: 9,
                },
            ),
            span: 1:20; [19, 27],
            typed_state: Untyped,
        },
        Expression {
            id: ExprId {
                id: 11,
            },
            kind: Assignment(
                IdentRefId {
                    id: 3,
                },
                ExprId {
                    id: 10,
                },
            ),
            span: 1:14; [13, 27],
            typed_state: Untyped,
        },
    ],
    statements: [
        Statement {
            id: StmtId {
                id: 0,
            },
            kind: Let(
                Identifier {
                    id: IdentId {
                        id: 0,
                    },
                    span: 1:5; [4, 7],
                },
                ExprId {
                    id: 0,
                },
            ),
            span: 1:1; [0, 12],
        },
        Statement {
            id: StmtId {
                id: 1,
            },
            kind: Expression(
                ExprId {
                    id: 5,
                },
            ),
            span: 1:14; [13, 28],
        },
        Statement {
            id: StmtId {
                id: 2,
            },
            kind: Let(
                Identifier {
                    id: IdentId {
                        id: 1,
                    },
                    span: 1:5; [4, 7],
                },
                ExprId {
                    id: 6,
                },
            ),
            span: 1:1; [0, 12],
        },
        Statement {
            id: StmtId {
                id: 3,
            },
            kind: Expression(
                ExprId {
                    id: 11,
                },
            ),
            span: 1:14; [13, 28],
        },
    ],
    root: [
        StmtId {
            id: 0,
        },
        StmtId {
            id: 1,
        },
        StmtId {
            id: 2,
        },
        StmtId {
            id: 3,
        },
    ],
    symbols: [],
    types: [],
    state: WithoutImplicitRet,
}
